### PR TITLE
fix(esp-modem): Add missing set_pdp_context() to C-API

### LINKED ISSUE #71: esp-modem build problems (IDFGH-7729)
IDF: 4.4.1
Verison: 0.1.17
Tool Chain: esp-2021r2-patch3-8.4.0
Target: esp32

```
Processing 3 dependencies:
[1/3] espressif/esp-modbus (1.0.3)
[2/3] espressif/esp_modem (0.1.17)
[3/3] idf (4.4.1)
```

Hi we are attempting to migrate from the older version of the esp-modem to the latest version and have come across some build issues.

#### Unknown netif type
```
In file included from /Users/ash/dev/project/managed_components/espressif__esp_modem/include/esp_modem_api.h:19,
                 from /Users/ash/dev/project/components/cellular-test/src/cellular_test.c:11:
/Users/ash/dev/project/managed_components/espressif__esp_modem/include/esp_modem_c_api_types.h:66:116: error: unknown type name 'esp_netif_t'; did you mean 'esp_err_t'?
 esp_modem_dce_t *esp_modem_new(const esp_modem_dte_config_t *dte_config, const esp_modem_dce_config_t *dce_config, esp_netif_t *netif);
                                                                                                                    ^~~~~~~~~~~
                                                                                                                    esp_err_t
```

#### Invalid cast

```
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c:180:10: warning: assignment to 'esp_modem_dce_t *' {aka 'struct esp_modem_dce_wrap *'} from 'int' makes pointer from integer without a cast [-Wint-conversion]
    p_dce = esp_modem_new_dev(ESP_MODEM_DCE_BG96, &dte_config, &dce_config, p_esp_netif);
```
Where p_dce is  'esp_modem_dce_t*'

### Missing functions
```
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c: In function 'cellular_connect_task':
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c:170:10: error: implicit declaration of function 'esp_modem_new_dev'; did you mean 'esp_modem_reset'? [-Werror=implicit-function-declaration]
  p_dce = esp_modem_new_dev(ESP_MODEM_DCE_BG96, &dte_config, &dce_config, p_esp_netif);
          ^~~~~~~~~~~~~~~~~
          esp_modem_reset
```

My cellular-test component has `esp_modem` in the PRIV_REQUIRES list, and I can see the component has been correctly configured in the managed_components directory.

### PR DESCRIPTION
**Pull request by david-cermak**
Closes #71

### CODE REVIEW

File: components/esp_modem/include/esp_modem_c_api_types.h [hunk 1]
```diff
@@ -22,7 +22,11 @@ extern "C" {
 
 typedef struct esp_modem_dce_wrap esp_modem_dce_t;
 
-struct PdpContext;
+struct PdpContext {
+    size_t context_id;
+    const char *protocol_type;
+    const char *apn;
+};
```
Comment (AshUK) [inline components/esp_modem/include/esp_modem_c_api_types.h:29]: a typedef here would be useful

Comment (david-cermak) [reply components/esp_modem/include/esp_modem_c_api_types.h:29]: Thanks for the feedback, will add

File: components/esp_modem/src/esp_modem_c_api.cpp [hunk 2]
```diff
@@ -393,3 +393,14 @@ extern "C" esp_err_t esp_modem_set_gnss_power_mode(esp_modem_dce_t *dce_wrap, in
...
+
+extern "C" esp_err_t esp_modem_set_pdp_context(esp_modem_dce_t *dce_wrap, struct ::PdpContext *c_api_pdp)
+{
+    if (dce_wrap == nullptr || dce_wrap->dce == nullptr) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_modem::PdpContext pdp{c_api_pdp->apn};
+    pdp.context_id = c_api_pdp->context_id;
+    pdp.protocol_type = c_api_pdp->protocol_type;
+    return command_response_to_esp_err(dce_wrap->dce->set_pdp_context(pdp));
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_modem/src/esp_modem_c_api.cpp:405]: Do you think that setting pdp context needs such a big timeout? It usually gets updated immediately on my device. Will check the spec...

[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_modem/src/esp_modem_c_api.cpp:405]: Checked that the timeout to define the *pdp-context* is about 300ms. Closing this discussion, feel free to reopen.

**Review by euripedesrocha — APPROVED** [review]

**Pull request merged**